rfcs/0148-pipe-operator.md

```diff
@@ -0,0 +1,327 @@
+`lib.pipe` may be the most powerful library function with that regard,
+but it is unknown and overlooked by many because it is not easily discoverable:
+Despite its great usefulness, it is currently used in less than 30 files in Nixpkgs
+(`rg '[\. ]pipe .* \['`).
+Additionally, it is not accessible to Nix code outside of nixpkgs,
+and due to Nix's lazy evaluation debugging type errors is really difficult.
+
+Let's have a look at an arbitrarily chosen snippet of Nixpkgs code:
+
+```nix
+defaultPrefsFile = pkgs.writeText "nixos-default-prefs.js" (lib.concatStringsSep "\n" (lib.mapAttrsToList (key: value: ''
+  // ${value.reason}
+  pref("${key}", ${builtins.toJSON value.value});
+'') defaultPrefs));
+```
+
+It is arguably pretty hard to read and reason about. Even when applying some more whitespace-generous formatting:
+
+```nix
+defaultPrefsFile = pkgs.writeText "nixos-default-prefs.js" (
+  lib.concatStringsSep "\n" (
+    lib.mapAttrsToList
+    (
+      key: value: ''
+        // ${value.reason}
+        pref("${key}", ${builtins.toJSON value.value});
+      ''
+    )
+    defaultPrefs
+  )
+);
+```
+
+One can observe the following issues:
+
+- If you want to follow the data flow, you must read it from bottom to top,
+  from the inside to the outside (the input here is `defaultPrefs`).
+- Adding a function call to the output would require wrapping the entire
+  expression in parentheses and increasing its indentation.
+
+Compare this to the equivalent call with `lib.pipe`:
+
+```nix
+defaultPrefsFile = pipe defaultPrefs [
+  (lib.mapAttrsToList (
+    key: value: ''
+      // ${value.reason}
+      pref("${key}", ${builtins.toJSON value.value});
+    ''
+  ))
+  (lib.concatStringsSep "\n")
+  (pkgs.writeText "nixos-default-prefs.js")
+];
+```
+
+The code now clearly reads from top to bottom in the order the data is processed,
+it is easy to add and remove processing steps at any point.
+
+With a dedicated pipe operator, it would look like this:
+
+```nix
+defaultPrefsFile = defaultPrefs
+  |> lib.mapAttrsToList (
+    key: value: ''
+      // ${value.reason}
+      pref("${key}", ${builtins.toJSON value.value});
+    ''
+  )
+  |> lib.concatStringsSep "\n"
+  |> pkgs.writeText "nixos-default-prefs.js";
+```
+
+The artificial distinction between the first input and the functions via the list now is gone,
+and so are the parentheses around the functions.
+With the lower character overhead, using the operator becomes attractive in more situations,
+whereas a `pipe` pays for its overhead only in more complex scenarios (usually three functions or more).
+Having a dedicated operator also increases visibility and discoverability of the feature.
+
+# Detailed design
+[design]: #detailed-design
+
+## `|>` operator
+
+A new operator `|>` is introduced into the Nix language.
+It is defined as function application with the order of arguments swapped: `f a` = `a |> f`.
+It is left-associative and has a binding strength one weaker than function application:
```

Comment:
Currently we are parsing "function application" like this"

```
expr_op : '!' expr_op
        | '-' expr_op
        | expr_op BINARY_OP expr_op
        | expr_app

%right ->
%left ||
%left &&
%nonassoc == !=
%nonassoc < > <= >=
%right //
%left NOT
%left + -
%left * /
%right ++
%nonassoc '?'
%nonassoc NEGATE

expr_app : expr_app expr_select
         | expr_select
```

So what it's binding strength in comparison to other binary operators?


(i.e. how should we interpret this expression)

```
1 + 2 |> 3
```

Reply:
IIUC, it'd be equivalent to `1 + (2 |> 3)`.

Without a type error, the example could be:
`1 + 2 |> f` = `1 + (2 |> f)` = `1 + f 2`

Reply:
> IIUC, it'd be equivalent to `1 + (2 |> 3)`.
> 
> Without a type error, the example could be: `1 + 2 |> f` = `1 + (2 |> f)` = `1 + f 2`

Why not `(1 + 2) > 3` ? I think we should explicitly state the relationship in this RFC

Reply:
> > IIUC, it'd be equivalent to `1 + (2 |> 3)`.
> > Without a type error, the example could be: `1 + 2 |> f` = `1 + (2 |> f)` = `1 + f 2`
> 
> Why not `(1 + 2) > 3` ? I think we should explicitly state the relationship in this RFC. (i.e. the _precedence_ among other binary operators.)

100% agree.

I'm not a javascript person at all but they're spending a lot of time figuring out exactly this over at https://github.com/tc39/proposal-pipeline-operator . That proposal is for hack-style pipe operators, and they discuss the difference with so-called "F# style" pipe operators, which I think is what this here RFC 0148 is going for, right? See e.g. https://github.com/valtech-nyc/proposal-fsharp-pipelines?tab=readme-ov-file , particularly the links in that README. It could be worth picking a specific language (F#?) and imitating their syntax as closely as possible. This means we can reuse all their hard work in figuring out the ergonomics, and it makes it an easier sell to the community :) "Nix now supports F# pipes".

Reply:
How about telling me where to insert "pipe" into this table. It specifies the precedence of all operators and should be updated accordingly once the pipe operator is introduced: https://nixos.org/manual/nix/stable/language/operators

Reply:
@inclyc 

> Why not (1 + 2) > 3

The plus operator (+) has precedence of 7. While function application has precedence of 2.
If the `|>` operator had weaker strength than (+) this opens a whole new dimension of complexity. It would allow many binary/unary operations. (i.e., all those from 1 to 7). Giving function application and pipe operator neighboring precedence means they behave similarly which I think is good behavior since they do similar things.

Reply:
"a binding strength one weaker than function application." yeah seeing this in context, I think this is just wrong.

We also had the discussion about the binding strength and associativity of `|>` and `<|`, and came to the preliminary conclusion that it should be forbidden to mix both on the same level. Maybe we could expand this and restrict the expressions on either side of `|>` to not involve *any* other operators?

Reply:
Also, forbidden constructs can be allowed in the future, whereas mistakenly allowed constructs require indefinite support unless we give up on long term reproducibility.

I would expect the pipe operators to form somewhat large expressions, where an extra pair of parentheses doesn't really matter much, and helps with clarity anyway.

So I agree that this is probably a good way to reduce scope and certainly reduce risk, although it *might* make the grammar more complicated and cause somewhat of a less obvious usage of some parser libraries/generators.

Reply:
Shepherds currently believe this conversation is resolved, outcome: several options are viable here, and an experimental implementation is probably more useful than more talk for figuring out which is best.